src/engagement_db_coda_sync/lib.py

```diff
@@ -1,15 +1,106 @@
+
+    default_user_ids = []
+    if not all_datasets_have_user_file_url:
+        assert coda_config.project_users_file_url is not None, \
+         f"Specify user_ids for coda datasets in CodaDatasetConfiguration or user_ids for this project in CodaSyncConfiguration"
+        default_user_ids = get_coda_users_from_gcloud(coda_config.project_users_file_url, google_cloud_credentials_file_path)
+
+    ws_correct_dataset_code_scheme = coda_config.ws_correct_dataset_code_scheme
+    for dataset_config in coda_config.dataset_configurations:
+        user_ids = []
+        if dataset_config.coda_dataset_users_file_url:
+            user_ids.append(get_coda_users_from_gcloud(dataset_config.coda_dataset_users_file_url))
+        else:
+            user_ids = default_user_ids
+
+        coda_user_ids = coda.get_dataset_user_ids(dataset_config.coda_dataset_id)
+        if coda_user_ids:
+            user_ids = list(set(user_ids) - set(coda_user_ids))
+            coda.update_dataset_user_ids(dataset_config.coda_dataset_id, user_ids)
+        else:
+            coda.set_dataset_user_ids(dataset_config.coda_dataset_id, user_ids)
+
+        repo_code_schemes = []
+        for code_scheme_config in dataset_config.code_scheme_configurations:
+            assert code_scheme_config.coda_code_schemes_count < 5, \
+                "`code_scheme_configuration.coda_code_schemes_count` exceeds number of columns that can be displayed in coda"
+            for count in range(1, code_scheme_config.coda_code_schemes_count + 1):
+                if count == 1:
+                    repo_code_schemes.append(code_scheme_config.code_scheme)
+                else:
+                    code_scheme_copy = CodeScheme.from_firebase_map(code_scheme_config.code_scheme.to_firebase_map())
+                    code_scheme_copy.scheme_id = f"{code_scheme_copy.scheme_id}-{count}"
+                    repo_code_schemes.append(code_scheme_copy)
+        repo_code_schemes.append(ws_correct_dataset_code_scheme)
+        repo_code_schemes_lut = {code_scheme.scheme_id: code_scheme for code_scheme in repo_code_schemes}
+
+        coda_code_schemes = coda.get_all_code_schemes(dataset_config.coda_dataset_id)
+        coda_scheme_ids = [code_scheme.scheme_id for code_scheme in coda_code_schemes]
+
+        if len(repo_code_schemes) != len(coda_code_schemes):
+            for scheme_id in coda_scheme_ids:
+                if scheme_id not in repo_code_schemes_lut.keys():
+                    log.warning(f"There are code schemes in coda not in this repo; The code schemes will be ignored")
+                    coda_code_schemes.remove(repo_code_schemes_lut[scheme_id])
+
+            for scheme_id, code_scheme in repo_code_schemes_lut.items():
+                if scheme_id not in coda_scheme_ids:
+                    coda.set_dataset_code_scheme(dataset_config.coda_dataset_id, code_scheme)
+                    repo_code_schemes.remove(code_scheme)
```

Comment:
I think you want to run this check even if len(repo_code_schemes) == len(coda_code_schemes), because the underlying schemes could be different even if the lengths are the same

Reply:
yeah, thanks. updated.